### ISSUE TITLE
Converters applied to Field.like

### ISSUE THREAD

**Issue by pschichtel** [jOOQ/jOOQ#12472]
### Expected behavior

I'm working with phone numbers and I have various wrappers for different kinds of phone numbers. I just introduced stricter validation during the creation of these wrapper types. This worked mostly fine except for a few queries that use the like comparator, which fails because e.g. `+49%` (search for numbers by prefix) is not a valid number. My expectation was, that the like operator, given that it only operates on strings anyway, would not apply type conversions.

### Actual behavior

Like operator applies type conversions which can lead to broken type validations, which leads to validation errors.

### Steps to reproduce the problem

1. Have string column with a custom type
2. Implement a converter that throws an exception if the string constains `%`
3. Use the like comparator on the column with e.g. `a%`.

### Versions

- jOOQ: 3.15.3
- Java: 16.0.2
- Database (include vendor): PostgreSQL
- OS: Fedora
- JDBC Driver (include name if inofficial driver): postgresql:42.2.23

**Comment by lukaseder** [comment]
> My expectation was, that the like operator, given that it only operates on strings anyway, would not apply type conversions.

Why would it not? There are reasonable use-cases for `Converter<String, String>`. I don't see why such a converter wouldn't apply to `LIKE`. In fact, I have often wondered if accepting a `Field<String>` was a mistake, and we should accept `Field<T>`, instead.

Of course, both changes (the one you suggested, and the one I mentioned above) are backwards incompatible, and unlikely to make it into jOOQ soon. Nevertheless, I'll leave this open for now. There are numerous ideas on this issue tracker about how to improve things with respect to the usual `<T, U>` type pairs, see https://github.com/jOOQ/jOOQ/issues/11088.

A workaround for you is to use plain SQL templating via `DSL.field("?", String.class, pattern)`, which will prevent the converter from being applied.

**Comment by pschichtel** [comment]
valid point with String->String conversions.

I've added a couple of extensions based on your suggestion:

```kotlin
fun <T> Field<T>.likeRaw(pattern: String) = like(unconvertedStringField(pattern))
fun <T> Field<T>.notLikeRaw(pattern: String) = notLike(unconvertedStringField(pattern))
private fun unconvertedStringField(string: String): Field<String> = field("?", String::class.java, string)
```

**Comment by lukaseder** [comment]
I've been reviewing this a bit more and there are indeed two ways to think about this:

1. A regression has been introduced when https://github.com/jOOQ/jOOQ/issues/10438 was implemented in jOOQ 3.14. The solution would be to go back to previous behaviour.
2. The changes in the above issue weren't complete, and we should have improved `LIKE` to accept `Field<T>` as illustrated previously (see below quote). The solution would be to open up the API.

> Why would it not? There are reasonable use-cases for `Converter<String, String>`. I don't see why such a converter wouldn't apply to `LIKE`. In fact, I have often wondered if accepting a `Field<String>` was a mistake, and we should accept `Field<T>`, instead.

The question here is whether `LIKE` is really an binary operator similar to `=`, `<`, etc. where it is reasonable to expect converters from either side to apply to bind values on the other side (as per #10438, and similar improvements over the recent minor releases), or whether it is more like a function `LIKE(string, pattern)`, where the arguments (as well as their types) are independent of one another. Clearly, it isn't really a binary operator with potential properties like associativity, commutativity, distributivity, etc.

I tend to think that 1) is the right solution here. The fact that we use infix operator-like syntax is just hiding the fact that this is really a function (similar to the regular expression functions). Also, the `ESCAPE` clause, which acts as a modifier with yet another argument of type `char` shows that the funciton is overloaded:

- `LIKE(string, pattern)`
- `LIKE(string, pattern, escape)`

... and that the arguments are independently typed, not invariant as with binary operators.

**Comment by pschichtel** [comment]
I completely agree with the interpretation that LIKE is not a binary operator, but a function.
